server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java

```diff
@@ -339,8 +340,19 @@ public void writeTo(StreamOutput out) throws IOException {
             }
         } else {
             // an old node will only understand legacy roles since pluggable roles is a new concept
-            final List<DiscoveryNodeRole> rolesToWrite =
-                    roles.stream().filter(DiscoveryNodeRole.LEGACY_ROLES::contains).collect(Collectors.toList());
+            final List<DiscoveryNodeRole> rolesToWrite = new ArrayList<>(roles.size());
+            boolean foundDataContainingRole = false;
+            for (DiscoveryNodeRole role : roles) {
+                if (DiscoveryNodeRole.LEGACY_ROLES.contains(role)) {
+                    rolesToWrite.add(role);
+                } else if (foundDataContainingRole == false && role.canContainData()) {
+                    // we only want to write out one "data" role even if the node has multiple data tier node roles
+                    // (eg. data_hot and data_content)
+                    foundDataContainingRole = true;
+                    rolesToWrite.add(DiscoveryNodeRole.DATA_ROLE);
+                }
+            }
+
```

Comment:
This is implemented differently than I was originally thinking, what about something like:

```suggestion
final Set<DiscoveryNodeRole> rolesToWrite = roles.stream()
                    .map(DiscoveryNodeRole::getCompatibilityRole)
                    .filter(DiscoveryNodeRole.LEGACY_ROLES::contains)
                    .collect(Collectors.toSet());
```

Which I think takes care of it, since these have compatibility roles that are the `DATA` role.

(I changed it to a set to avoid duplicates, and also I didn't test that this compiles but hopefully you get the idea)

Reply:
Great suggestion! Thanks @dakrone